contracts/utils/Swap.sol

```diff
@@ -0,0 +1,454 @@
+     * @dev The nonce states are encoded as 256 bits, for each nonce in the group 0 means available and 1 means used
+     */
+    mapping(address => mapping(uint256 => uint256)) internal _nonceGroups;
+
+    /************************************************
+     *  CONSTRUCTOR
+     ***********************************************/
+
+    constructor() {
+        uint256 currentChainId = getChainId();
+        DOMAIN_CHAIN_ID = currentChainId;
+        DOMAIN_SEPARATOR = keccak256(
+            abi.encode(
+                DOMAIN_TYPEHASH,
+                DOMAIN_NAME,
+                DOMAIN_VERSION,
+                currentChainId,
+                this
+            )
+        );
+    }
+
+    /************************************************
+     *  SETTER
+     ***********************************************/
+
+    /**
+     * @notice Sets the referral fee for a specific referrer
+     * @param referrer is the address of the referrer
+     * @param fee is the fee in percent in 2 decimals
+     */
+    function setFee(address referrer, uint256 fee) external onlyOwner {
+        require(referrer != address(0), "Referrer cannot be the zero address");
+        require(fee < MAX_PERCENTAGE, "Fee exceeds maximum");
+
+        referralFees[referrer] = fee;
+    }
+
+    /************************************************
+     *  OFFER CREATION AND SETTLEMENT
+     ***********************************************/
+
+    /**
+     * @notice Create a new offer available for swap
+     * @param offeredToken token offered by seller
+     * @param biddingToken token asked by seller
+     * @param minPrice minimum price of offeredToken in terms of biddingToken
+     * @param minBidSize minimum size allowed in terms of biddingToken
+     * @param totalSize amount of offeredToken offered by seller
+     */
+    function createOffer(
+        address offeredToken,
+        address biddingToken,
+        uint96 minPrice,
+        uint96 minBidSize,
+        uint128 totalSize
+    ) external override returns (uint256 swapId) {
+        require(
+            offeredToken != address(0),
+            "OfferedToken cannot be the zero address"
+        );
+        require(
+            biddingToken != address(0),
+            "BiddingToken cannot be the zero address"
+        );
+        require(minPrice > 0, "MinPrice must be larger than zero");
+        require(minBidSize > 0, "MinBidSize must be larger than zero");
+        require(totalSize > 0, "TotalSize must be larger than zero");
+
+        offersCounter += 1;
+
+        swapId = offersCounter;
+
+        swapOffers[swapId].seller = msg.sender;
+        swapOffers[swapId].isOpen = true;
+        swapOffers[swapId].offeredToken = offeredToken;
+        swapOffers[swapId].biddingToken = biddingToken;
+        swapOffers[swapId].offeredTokenDecimals = IERC20Detailed(offeredToken)
+            .decimals();
```

Comment:
Since we are just offering otokens, shall we just delete `offeredTokenDecimals` from the struct entirely? We can hardcode it to 8 decimals in the swap